src/lib.rs

```diff
@@ -1,68 +1,73 @@
-#![cfg_attr(feature = "nightly", deny(missing_docs))]
-#![cfg_attr(feature = "nightly", feature(external_doc))]
-#![cfg_attr(feature = "nightly", doc(include = "../README.md"))]
-#![cfg_attr(test, deny(warnings))]
-
-/// Returns how many bytes are needed to encode a value.
-#[inline]
-pub fn length(value: u64) -> usize {
-  let zero_len = 64 - value.leading_zeros();
-  let offset = if zero_len == 0 { 7 } else { 6 };
-  ((offset + zero_len) / 7) as usize
-}
+extern crate failure;
+use failure::{bail, Error};
 
 /// Encode a `u64` integer to the byte slice. Returns how many bytes were
 /// encoded.
 #[inline]
-pub fn encode(value: u64, buf: &mut [u8]) -> usize {
+pub fn encode(value: u64, buf: &mut [u8]) -> Result<usize, Error> {
   encode_with_offset(value, buf, 0)
 }
 
 /// Encode a `u64` integer at a specific offset in the byte slice. Returns how
 /// many bytes were encoded.
 #[inline]
-pub fn encode_with_offset(value: u64, buf: &mut [u8], offset: usize) -> usize {
+pub fn encode_with_offset(
+  value: u64,
+  buf: &mut [u8],
+  offset: usize,
+) -> Result<usize, Error> {
+  let len = length(value);
+  if buf.len() < len {
+    bail!["buffer is too small to write varint"]
```

Comment:
When not using `failure` anymore, this would just be `return DecodeError`, and the message would be in the `Display` impl of `DecodeError` (see [here](https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/define_error_type.html) for an example).